### PR TITLE
[BUG/FIX] Azure RBAC fix refresh token logic

### PR DESCRIPTION
**Pull request by saisankargochhayat**
This PR aims to fix the refresh token logic for Azure RBAC.
The existing refresh token logic relies on the `expires_in` value from the token response -
{"access_token":"<toeken>","expires_in":"86700","refresh_token":"","expires_on":"1732881796","not_before":"1732795096","resource":"https://management.azure.com","token_type":"Bearer"}
However in the case when the token is not fresh and returned from cache, the `expires_in` value misleads Guard to think the token doesn't need refreshing since it adds the current_time + expires_in to set the new expiry time.
In this PR we directly use the `expires_on` to set the expiry time accurately.

The PR also increases the tokenExpiryDelta to 5 minutes to ensure the token is refreshed 5 mins earlier than actual expiry. This is done on both AuthZ (rbac.go) and AuthN (graph.go) flow.

Removed `expires_in` to be returned from msi_adapter in favor of `expires_on`

Using the current implementation guard now sets the expires property correctly, i.e. knows it expires at 2024-11-29 21:05:56 +0000 UTC in the below example rather than doing - current-time + 86700 

Guard log - 
```
│ I1202 20:38:40.019063       1 graph.go:331] Token received with expires_in 86700 and expires_at 1733258179            

││ I1202 20:38:40.019114       1 graph.go:337] Token refreshed successfully at 2024-12-02 20:38:40.019110448 +0000 UTC m=+444.823514368. Expire at set to: 2024-12-03 20:31:19 +0000 UTC 
 ```
 
 Tested KAP flow - 
 
image

RBAC non proxy flow - 
image

### CODE REVIEW

File: auth/providers/azure/graph/graph.go
Comment (vineeth-thumma) [inline auth/providers/azure/graph/graph.go:75]: rename to `tokenExpiryDelta` to be specific?

**Review by vineeth-thumma — CHANGES_REQUESTED** [review]

File: auth/providers/azure/graph/graph.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph.go:75]: This is not a new var, but I can add a comment?

File: auth/providers/azure/graph/graph.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph.go]: Updated to at

File: auth/providers/azure/graph/types.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/types.go]: Renamed to ExpiresIn

File: auth/providers/azure/graph/graph.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph.go]: Updated to suggested name

File: auth/providers/azure/graph/graph.go
Comment (AzureMarker) [inline auth/providers/azure/graph/graph.go:76]: Can you also call out this change from 1min to 5min in the PR description?

File: auth/providers/azure/graph/graph_test.go
Comment (AzureMarker) [inline auth/providers/azure/graph/graph_test.go:114]: Can you add a check to make sure the expiration time is exactly the `expiresOn` value minus 5min?
Currently the test doesn't validate the change in behavior.

File: auth/providers/azure/graph/msi_tokenprovider_test.go
Comment (AzureMarker) [inline auth/providers/azure/graph/msi_tokenprovider_test.go:72]: To simplify this, can you have the expected `expires_on` value be a constant that gets used in the mock token response and here?

File: authz/providers/azure/rbac/rbac.go
Comment (AzureMarker) [inline authz/providers/azure/rbac/rbac.go:59]: Also call out that the time delta changed for authz.

File: authz/providers/azure/rbac/rbac_test.go
Comment (AzureMarker) [inline authz/providers/azure/rbac/rbac_test.go:185]: This test should validate the new behavior.

File: auth/providers/azure/graph/graph.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph.go:76]: Updated

File: authz/providers/azure/rbac/rbac.go
Comment (saisankargochhayat) [reply authz/providers/azure/rbac/rbac.go:59]: Updated

File: auth/providers/azure/graph/msi_tokenprovider_test.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/msi_tokenprovider_test.go:64]: Updated based on suggestion

File: auth/providers/azure/graph/graph_test.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph_test.go]: Updated to test the expected behavior

File: authz/providers/azure/rbac/rbac_test.go
Comment (saisankargochhayat) [reply authz/providers/azure/rbac/rbac_test.go]: Updated to test the expected behavior

File: auth/providers/azure/graph/msi_tokenprovider_test.go
Comment (AzureMarker) [inline auth/providers/azure/graph/msi_tokenprovider_test.go:31]: I think you put the `%d` on the wrong property?

Comment (saisankargochhayat) [reply auth/providers/azure/graph/msi_tokenprovider_test.go]: Nice catch thanks!

File: auth/providers/azure/graph/msi_tokenprovider.go
Comment (julienstroheker) [inline auth/providers/azure/graph/msi_tokenprovider.go:100]: IMO if we don't use this, we should remove it from the parsed responsed json payload to avoid using it elsewhere

File: auth/providers/azure/graph/types.go
Comment (julienstroheker) [inline auth/providers/azure/graph/types.go:22]: Like mentioned we should remove this

**Review by julienstroheker — COMMENTED** [review]
LGTM, I'll let you decide if you want to apply my comment about the returned payload from MSI adapter

**Comment by julienstroheker** [comment]
Also rename the PR with [BUG/FIX] in the title

File: auth/providers/azure/graph/graph_test.go
Comment (julienstroheker) [inline auth/providers/azure/graph/graph_test.go:110]: you should then update this paylod to remove the `expired_in` and anywhere else

Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph_test.go:109]: Updated

File: auth/providers/azure/graph/types.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/types.go]: Updated

**Comment by vineeth-thumma** [comment]
LGTM!

File: auth/providers/azure/graph/graph_test.go
Comment (saisankargochhayat) [reply auth/providers/azure/graph/graph_test.go]: Thanks fixed

File: authz/providers/azure/rbac/rbac_test.go
Comment (saisankargochhayat) [reply authz/providers/azure/rbac/rbac_test.go]: Thanks fixed

**Review by julienstroheker — APPROVED** [review]
LGTM on Arc Side but need review on AKS Side since we are touching the token interface.

File: auth/providers/azure/graph/graph_test.go
Comment (vineeth-thumma) [inline auth/providers/azure/graph/graph_test.go:128]: this comment is not relevant anymore? or might be confusing given that you're normalizing it above for comparision

File: authz/providers/azure/rbac/rbac_test.go
Comment (vineeth-thumma) [inline authz/providers/azure/rbac/rbac_test.go:202]: this comment is not relevant anymore this comment is not relevant anymore? or might be confusing given that you're normalizing it above for comparison

**Review by vineeth-thumma — CHANGES_REQUESTED** [review]

**Review by vineeth-thumma — APPROVED** [review]
LGTM

File: authz/providers/azure/rbac/rbac.go
Comment (bcho) [inline authz/providers/azure/rbac/rbac.go:231]: Will the tokenProvider.Acquire guarantee to refresh the token or it's possible to load from a cache? In the latter case, will there be chance that the `resp.ExpiresOn` is less than 5 mins from time.Now(), which causes the a.expiresAt < time.Now() ? What's the impact of seeing this, like will that cause the RefreshToken being call repeatedly and eventualy cause busy loop?

Comment (saisankargochhayat) [reply authz/providers/azure/rbac/rbac.go:231]: So, from the tokenProvider.Acquire point of view, it's always going to get a token with at-least 1 hour of expiry left
As the identity controller always refreshes the token 1 hour before expiry - 
https://github.com/azure-core/ClusterConfigurationAgent/blob/9e24b19ffdc7f71f53cb4fb86cb3d7d830be6856/Utils/TokenProviders/ManagedIdentityTokenProvider.go#L78

So, we should never hit a scenario where the ExpiresOn is less than 5mins from time.Now()

**Review by bcho — APPROVED** [review]

File: authz/providers/azure/rbac/rbac.go
Comment (julienstroheker) [reply authz/providers/azure/rbac/rbac.go:231]: For context, in Arc we have on our own `msi-adapter` component that mimic IMDS on the edge. Like @saisankargochhayat it just fetch a "cached" token from another component "identity-controller" that manage getting token and save it as k8s secret via a CR. The code that @saisankargochhayat linked is where `msi-adapter` decide to trigger a token rotation of re-used the cache secret within the 1H expiration.

**Review by julienstroheker — APPROVED** [review]

**Review by weinong — APPROVED** [review]
LGTM

**Pull request merged**
